libs/metrics/test_positivity.py

```diff
@@ -80,41 +114,64 @@ def _make_output_dataset(
-    #  Something like:
-    # assert dataset.provenance.index.names == [CommonFields.LOCATION_ID, PdFields.VARIABLE]
-    # provenance = dataset.provenance.loc[locations, self._column]
     assert dataset_in.provenance.index.names == [CommonFields.LOCATION_ID, PdFields.VARIABLE]
+    if source_columns:
+        # Make a DataFrame of the provenance with location_id index and column for each of
+        # source_columns. The columns are in an arbitrary order.
+        provenance_unsorted_columns = dataset_in.provenance.loc[locations, source_columns].unstack(
+            PdFields.VARIABLE
+        )
+        provenance = (
+            provenance_unsorted_columns
+            # TODO(tom): remove the hack that derives provenance from the field name and use
+            #  something like the following to produce a more complete provenance per location.
+            #  Maybe make this a method of the class Method and use self.columns instead of
+            #  source_columns.
+            # .apply(lambda s: f"{default_provenance}({','.join(s)})", axis=1)
+            .apply(lambda s: ",".join(s.drop_duplicates()), axis=1).rename(PdFields.PROVENANCE)
+        )
+        _append_variable_index_level(provenance, output_metric)
+    else:
+        provenance = pd.Series([], name=PdFields.PROVENANCE, dtype="object")
 
     # Make sure every timeseries has something in `provenance`.
     provenance_index = pd.MultiIndex.from_product(
         [locations, [output_metric]], names=[CommonFields.LOCATION_ID, PdFields.VARIABLE],
     )
-    provenance = pd.Series([], name=PdFields.PROVENANCE, dtype="object").reindex(
-        provenance_index, fill_value=default_provenance
-    )
+    provenance = provenance.reindex(provenance_index, fill_value=default_provenance)
     return MultiRegionDataset.from_timeseries_wide_dates_df(wide_date_df).add_provenance_series(
         provenance
     )
 
 
 @dataclasses.dataclass
-class DivisionMethod(Method):
-    """A method of calculating test positivity by dividing a numerator by a denominator"""
+class _DivisionMethodAttributes:
+    """Non-default attributes of DivisionMethod, extracted to avoid TypeError.
 
-    _name: str
+    The problem and this somewhat ugly solution are described at
+    https://stackoverflow.com/q/51575931
+    """
+
+    _name: timeseries.DatasetName
     _numerator: FieldName
     _denominator: FieldName
 
+
+@dataclasses.dataclass
+class DivisionMethod(Method, _DivisionMethodAttributes):
+    """A method of calculating test positivity by dividing a numerator by a denominator"""
+
     @property
-    def name(self) -> str:
+    def name(self) -> timeseries.DatasetName:
         return self._name
 
     @property
     def columns(self) -> Set[FieldName]:
         return {self._numerator, self._denominator}
 
-    def calculate(self, dataset: MultiRegionDataset, diff_days: int) -> MultiRegionDataset:
+    def calculate(
+        self, dataset: MultiRegionDataset, diff_days: int, most_recent_date
```

Comment:
type on most_recent_date?

Reply:
added pd.Timestamp, which seems to work and I think is right most of the time, though maybe str works too.